### PR TITLE
Add 'typeof' operator to DataPrepper expressions

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
[Describe what this change achieves]
 
### Issues Resolved
Resolves #[Issue number to be closed when this PR is merged]
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by graytaylor0 — DISMISSED** [review]
Looks good. This is in draft mode though. Also please add a sample yaml usage in the PR details

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java [hunk 1]
```diff
@@ -75,4 +75,22 @@ public String getTypeName() {
     static DataType fromTypeName(final String option) {
         return TYPES_MAP.get(option);
     }
+
+    public static boolean isSameType(final Object object, final String option) {
+        DataType type = fromTypeName(option);
+        if (type == null)
```
Comment (dinujoh) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java:81]: We should have a check for object is null.

`/myfield typeof null`

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java [hunk 2]
```diff
@@ -75,4 +75,22 @@ public String getTypeName() {
...
+            throw new IllegalArgumentException("Unknown DataType");
+        switch (type) {
```
Comment (dinujoh) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java:83]: we should add a case condition for BigDecimal.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java (hunk 1 above)
Comment (kkondaka) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java:81]: I do not think that's necessary. We already support `/myfield == null`. I do not think `typeof null` makes sense.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java [hunk 3]
```diff
@@ -75,4 +75,22 @@ public String getTypeName() {
...
+          case STRING:
+            return (object instanceof String);
+          case BOOLEAN:
+            return (object instanceof Boolean);
+          case INTEGER:
+            return (object instanceof Integer);
+          case LONG:
+            return (object instanceof Long);
+          default: // DOUBLE
+            return (object instanceof Double);
```
[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java:93]: Is this required for docdb release or for some future use case?

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/DataTypeTest.java [hunk 4]
```diff
@@ -5,16 +5,46 @@
...
-}
\ No newline at end of file
+
+    @Test
+    void test_isSameType_withInvalidInput() {
+        assertThrows(IllegalArgumentException.class, () -> DataType.isSameType(2, "unknown"));
+    }
+
+    @Test
+    void test_isSameType() {
```
Comment (dinujoh) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/DataTypeTest.java:31]: Consider adding parameterized test to cover all cases.

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/TypeOfOperatorTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        assertThat(objectUnderTest.shouldEvaluate(ctx), is(false));
+    }
+
+    @Test
+    void testGetSymbol() {
+        assertThat(objectUnderTest.getSymbol(), is(DataPrepperExpressionParser.TYPEOF));
+    }
+
+    @Test
+    void testEvalValidArgs() {
```
Comment (dinujoh) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/TypeOfOperatorTest.java:48]: this case be parameterized test as well.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java (hunk 2 above)
Comment (kkondaka) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/DataType.java:97]: After discussion, decided not to support these two for now as double handles float and all BigDecimals are actually strings.

**Review by dinujoh — APPROVED** [review]

**Review by graytaylor0 — APPROVED** [review]

**Pull request merged**